internals/overlord/state/state.go

```diff
@@ -570,6 +568,131 @@ NextChange:
 			delete(s.tasks, tid)
 		}
 	}
+	// Prune expired notices, and update the latest warning time cache.
+	var latestWarningTime time.Time
+	for k, n := range s.notices {
+		if n.expired(now) {
+			stats.IncludeNotice(n)
+			delete(s.notices, k)
+		} else if n.noticeType == WarningNotice {
+			if n.lastRepeated.After(latestWarningTime) {
+				latestWarningTime = n.lastRepeated
+			}
+		} else if n.noticeType == ChangeUpdateNotice {
+			if _, exists := s.changes[n.key]; !exists {
+				stats.IncludeNotice(n)
+				delete(s.notices, k)
+			}
+		}
+	}
+	s.latestWarningTime.Store(&latestWarningTime)
+	if len(s.notices) > maxNotices {
+		s.pruneMaxNotices(maxNotices, stats)
+	}
+	stats.Log()
+}
+
+// pruneStats tracks how many changes and notices have been pruned, and what
+// time range has been affected.
+type pruneStats struct {
+	numChangesPruned   int
+	numNoticesPruned   int
+	oldestChangePruned time.Time
+	newestChangePruned time.Time
+	oldestNoticePruned time.Time
+	newestNoticePruned time.Time
+}
+
+func (p *pruneStats) IncludeChange(chg *Change) {
+	p.numChangesPruned++
+	if chg == nil || chg.readyTime.IsZero() {
+		return
+	}
+	if p.oldestChangePruned.IsZero() || p.oldestChangePruned.After(chg.readyTime) {
+		p.oldestChangePruned = chg.readyTime
+	}
+	if p.newestChangePruned.IsZero() || p.newestChangePruned.Before(chg.readyTime) {
+		p.newestChangePruned = chg.readyTime
+	}
+}
+
+func (p *pruneStats) IncludeNotice(n *Notice) {
+	p.numNoticesPruned++
+	if n == nil || n.lastOccurred.IsZero() {
```

Comment:
Similar to above, let's remove `n == nil`. Also, `lastOccured` will always be set, so I think we should either remove this or panic if it's zero with "internal error: lastOccurred unexpectedly zero".